### PR TITLE
Setup ruff (v2)

### LINKED ISSUE #140: Setup Ruff
**Describe the feature you'd like**

Use Ruff https://docs.astral.sh/ruff/ to detect errors.

### PR DESCRIPTION
**Pull request by EarlMilktea**
**Description of the change:**

Added several important ruff rules.

**Related issue:**

Resolves #140 .

### CODE REVIEW

File: graphix/gflow.py [hunk 1]
```diff
@@ -872,114 +872,6 @@ def get_corrections_from_pattern(pattern: Pattern) -> tuple[dict[int, set[int]],
...
-            cand = find_odd_neighbor(G, corrections) - {node}
-            if cand:
-                zflow_from_xflow[node] = cand
-        if zflow_from_xflow != zflow:  # if zflow is consistent with xflow
-            return None, None
-        return xflow, l_k
-    else:
-        return None, None
-
-
```
Comment (shinich1) [inline graphix/gflow.py:982]: Note: these two were deleted because of duplicate definitions.

**Review by shinich1 — APPROVED** [review]
Looks good, thanks! @thierry-martinez please also take a look.

File: graphix/graphsim/rxgraphviews.py [hunk 2]
```diff
@@ -70,8 +81,14 @@ class EdgeList:
     """
 
     def __init__(
-        self, edge_nums: list[tuple[int, int]] = [], edge_datas: list[dict] = [], edge_indices: list[int] = []
+        self, edge_nums: list[tuple[int, int]] = None, edge_datas: list[dict] = None, edge_indices: list[int] = None
```
Comment (thierry-martinez) [inline graphix/graphsim/rxgraphviews.py:84]:
```suggestion
        self, edge_nums: list[tuple[int, int]] | None = None, edge_datas: list[dict] | None = None, edge_indices: list[int] | None = None
```

File: graphix/pattern.py [hunk 3]
```diff
@@ -62,10 +62,12 @@ class Pattern:
         total number of nodes in the resource state
     """
 
-    def __init__(self, input_nodes=[]):
+    def __init__(self, input_nodes=None):
```
Comment (thierry-martinez) [inline graphix/pattern.py:65]:
```suggestion
    def __init__(self, input_nodes: list[int] | None = None) -> None:
```

File: graphix/pattern.py [hunk 4]
```diff
@@ -2178,4 +2191,5 @@ def assert_permutation(original, user):
         if node in node_set:
             node_set.remove(node)
         else:
-            assert False, f"{node} appears twice"
+            msg = f"{node} appears twice"
+            raise ValueError(msg)
```
Comment (thierry-martinez) [inline graphix/pattern.py:2195]:
```suggestion
            raise ValueError(f"{node} appears twice")
```

**Review by thierry-martinez — CHANGES_REQUESTED** [review]
Some minor remarks, looks good otherwise. Thanks!

**Comment by EarlMilktea** [comment]
@thierry-martinez Updated. Thank you for suggestions!

**Review by thierry-martinez — APPROVED** [review]

**Comment by thierry-martinez** [comment]
Looks good to me now: ok for merge!

**Pull request merged**